File: validate.py
"""
身份证号校验器
"""
import numpy as np
import re
import datetime
from code2address import code2address


def get_check_code(x: str):
    assert len(x) == 17
    w = '07 09 10 05 08 04 02 01 06 03 07 09 10 05 08 04 02'.split()
    w = [int(i) for i in w]
    x = list(map(int, x))
    v = np.dot(x, w)
    check_code = (12 - v % 11) % 11
    return 'X' if check_code == 10 else str(check_code)


def valiate_check_code(x: str):
    check_code = get_check_code(x[:-1])
    if check_code != x[-1]:
        return f'校验位错误{v % 11}'


def validate_pattern(x: str):
    return re.match('\\d17[\\dX]', x)


def day_of(y, m):
    def is_leap(y):
        if y % 100 == 0:
            return y % 400 == 0
        else:
            return y % 4 == 0

    if m in (1, 3, 5, 7, 8, 10, 12):
        should = 31
    elif m == 2:
        should = 29 if is_leap(y) else 28
    else:
        should = 30
    return should


def validate_year_month_day(v: str):
    # 校验年月日
    y = int(v[6:10])
    m = int(v[10:12])
    d = int(v[12:14])
    now = datetime.datetime.now()

    if (y, m, d) > (now.year, now.month, now.day):
        return "不能在未来出生"
    if y < 1900:
        return "不可能如此长寿"
    if not 12 >= m >= 1:
        return "月份错误"

    if not day_of(y, m) >= d >= 1:
        return "日期错误"


def validate_address(x: str):
    # 校验地址
    address = x[:6]
    if address not in code2address:
        return '未知的地址'


def validate(x: str):
    for f in (validate_pattern,
              validate_year_month_day,
              valiate_check_code,
              validate_address):
        e = f(x)
        if e:
            return e


if __name__ == '__main__':
    x = 'xxx'
    print(validate(x))

File: idcard4rust/src/validate.rs
use crate::code2address::getBook;

/**
身份证号校验器
*/

fn validate_address(idcard: &String) -> String {
    let book = getBook();
    let code = &idcard[0..6];
    if book.contains_key(code) {
        "".to_string()
    } else {
        format!("没有省份{}", code)
    }
}

fn isLeapYear(year: i32) -> bool {
    if year % 100 == 0 {
        year % 400 == 0
    } else {
        year % 4 == 0
    }
}

pub fn getDaysOfMonth(year: i32, month: i32) -> i32 {
    if month == 2 {
        return if isLeapYear(year) { 29 } else { 28 };
    }
    if [1, 3, 5, 7, 8, 10, 12].contains(&month) {
        31
    } else {
        30
    }
}

fn validate_year_month_day(idcard: &String) -> String {
    //校验年月日
    fn parseNumber(s: &str) -> i32 {
        return s.parse::<i32>().unwrap();
    }
    let year: i32 = parseNumber(&idcard[6..10]);
    let month = parseNumber(&idcard[10..12]);
    let day = parseNumber(&idcard[12..14]);
    if year < 1920 || year > 2040 {
        return format!("年份错误{}", year);
    }
    if !(month >= 1 && month <= 12) {
        return format!("月份错误{}", month);
    }
    let days = getDaysOfMonth(year, month);
    if !(day >= 1 && day <= days) {
        return format!("日期错误{},总天数{}", day, days);
    }
    "".to_string()
}

fn idcard2int(idcard: &String) -> Vec<i32> {
    let x: Vec<i32> = idcard.chars().map(|x| {
        if x == 'X' { 10 } else { x as i32 - '0' as i32 }
    }).collect();
    x
}

const WEIGHT_STRING: &str = "07 09 10 05 08 04 02 01 06 03 07 09 10 05 08 04 02 01";

#[allow(non_upper_case_globals)]
pub fn getCheckCode(idcard: &String) -> char {
    //获取校验码
    let x = idcard2int(&idcard);
    static mut weight: Vec<i32> = Vec::new();
    //初始化weight
    unsafe {
        if weight.len() == 0 {
            weight = WEIGHT_STRING.split_whitespace().map(|x| {
                if x == "X" {
                    10
                } else {
                    x.parse::<i32>().unwrap()
                }
            }).collect();
            weight = weight[0..weight.len() - 1].to_owned();
        }
    }
    unsafe {
        if idcard.len() != weight.len() {
            panic!("idcard length is different with weight {} {}", idcard.len(), weight.len());
        }
        let mut s = 0u128;
        for i in 0..weight.len() {
            s += (x[i] * weight[i]) as u128;
        }
        let checkCode = (12 - s % 11) % 11;
        if checkCode == 10 { 'X' } else { ('0' as u128 + checkCode) as u8 as char }
    }
}

#[allow(non_upper_case_globals)]
fn validate_checkcode(idcard: &String) -> String {
    static mut weight: Vec<i32> = Vec::new();
    unsafe {
        if weight.len() == 0 {
            weight = WEIGHT_STRING.split_whitespace().map(|x| {
                if x == "X" {
                    return 10;
                }
                x.parse::<i32>().unwrap()
            }).collect();
        }
    }
    let nums: Vec<i32> = idcard2int(&idcard);
    if nums.len() != unsafe { weight.len() } {
        panic!("length should be equal")
    }
    let mut s = 0i128;
    unsafe {
        for i in 0..weight.len() {
            s += (weight[i] * nums[i]) as i128;
        }
    }

    if s % 11 == 1 {
        return "".to_string();
    }
    "checkcode error".to_string()
}

pub fn validate(idcard: &String) -> String {
    let validators = [validate_address, validate_year_month_day, validate_checkcode, ];
    for i in validators.iter() {
        let res: String = i(&idcard);
        if !res.is_empty() {
            return res;
        }
    }
    "".to_string()
}